docs/content/NavList.mdx

```diff
@@ -0,0 +1,536 @@
+```
+
+<details>
+  <summary>Rendered HTML</summary>
+
+```html
+<nav>
+  <ul role="list">
+    <li>
+      <a href="/inbox" aria-current="page">
+        <span>Inbox</span>
+        <span>1,234</span>
+      </a>
+    </li>
+    <li><a href="/saved">Saved</a></li>
+    <li><a href="/done">Done</a></li>
+  </ul>
+</nav>
+```
+
+</details>
+
+### With a heading
+
+```jsx
+<>
+  <h3 id="workflows-heading">Workflows</h3>
+  <NavList aria-labelledby="workflows-heading">
+    <NavList.Item href="/" current>
+      All workflows
+    </NavList.Item>
+    <NavList.Item href="/ci">CI</NavList.Item>
+    <NavList.Item href="/deploy">Deploy</NavList.Item>
+    <NavList.Item href="/release">Release</NavList.Item>
+  </NavList>
+</>
+```
+
+<details>
+  <summary>Rendered HTML</summary>
+
+```html
+<h3 id="workflows-heading">Workflows</h3>
+<nav aria-labelledby="workflows-heading">
+  <ul role="list">
+    <li><a href="/" aria-current="page">All workflows</a></li>
+    <li><a href="/ci">CI</a></li>
+    <li><a href="/deploy">Deploy</a></li>
+    <li><a href="/release">Release</a></li>
+  </ul>
+</nav>
+```
+
+</details>
+
+### With aria-label
+
+```jsx
+<NavList aria-label="Security">
+  <NavList.Item href="/" current>
+    Overview
+  </NavList.Item>
+  <NavList.Item href="/policy">Security policy</NavList.Item>
+  <NavList.Item href="/advisories">Security advisories</NavList.Item>
+</NavList>
+```
+
+<details>
+  <summary>Rendered HTML</summary>
+
+```html
+<nav aria-label="Security">
+  <ul role="list">
+    <li><a href="/" aria-current="page">Overview</a></li>
+    <li><a href="/policy">Security policy</a></li>
+    <li><a href="/advisories">Security advisories</a></li>
+  </ul>
+</nav>
+```
+
+</details>
+
+### With groups
+
+```jsx
+<NavList>
+  <NavList.Group title="Overview">
+    <NavList.Item href="/getting-started" current>
+      Getting started
+    </NavList.Item>
+  </NavList.Group>
+  <NavList.Group title="Components">
+    <NavList.Item href="/Avatar">Avatar</NavList.Item>
+    <NavList.Item href="/Button">Button</NavList.Item>
+    <NavList.Item href="/Label">Label</NavList.Item>
+  </NavList.Group>
+</NavList>
+```
+
+<details>
+  <summary>Rendered HTML</summary>
+
+```html
+<nav>
+  <ul role="list">
+    <li>
+      <div role="presentation" id="generated-id-1">Overview</div>
+      <ul role="list" aria-labelledby="generated-id-1">
+        <li><a href="/getting-started" aria-current="page">Getting started</a></li>
+      </ul>
+    </li>
+    <li>
+      <div role="presentation" id="generated-id-2">Components</div>
+      <ul role="list" aria-labelledby="generated-id-2">
+        <li><a href="/Avatar">Avatar</a></li>
+        <li><a href="/Button">Button</a></li>
+        <li><a href="/Label">Label</a></li>
+      </ul>
+    </li>
+  </ul>
+</nav>
+```
+
+</details>
+
+### With sub-items
+
+```jsx
+<NavList>
+  <NavList.Item href="/branches">Branches</NavList.Item>
+  <NavList.Item href="/tags">Tags</NavList.Item>
+  <NavList.Item>
```

Comment:
Should we give this item a special name? `NavList.Collapsable`, `NavList.Section`, `NavList.Accordion`, `NavList.Details`



```jsx
<NavList>
  <NavList.Item href="/branches">Branches</NavList.Item>
  <NavList.Item href="/tags">Tags</NavList.Item>
  <NavList.Collapsable title="Actions">
    <NavList.Item href="/branches">Branches</NavList.Item>
    <NavList.Item href="/tags">Tags</NavList.Item>
  </NavList.Collapsable/>
</NavList>
```

Section sounds nice but easy to confuse it with Group

Reply:
Looking at other primer implementations for inspiration,

primer/css uses [`.ActionList-item .ActionList-item--hasSubItem`](https://github.com/primer/css/blob/50221a5486d28c54dc1f578b459addd867bd01d3/src/actionlist/action-list-item.scss#L34), primer/view-components seems to [infers it from children](https://cs.github.com/github/github/blob/646833c1794e59114a06f223bf7dbe1d2ac3190f/app/components/primer/experimental/navigation_list.rb#L77)

Reply:
This is amazing feedback. Thank you @siddharthkp! I've scheduled some time to discuss the nuances of these API options. Feel free to move the calendar event if another time is more convenient :)

Another API consideration: TreeView/TreeGrid will need to support similar expand/collapse functionality. I would love to pick an API that we can share across all three components

Reply:
> /* assuming that's good, and subitem does not need to restricted to just text

Good question! @vdepizzol @langermank Should subitem support all the same functionality as item or is it more restrictive?

Reply:
It's kind of a long story, _but_, I refactored `NavList` to use semantic `<button>` elements for the collapse items (which originally had a click handler on `<li>`) allowing us to remove a bunch of JS. After I made that same change to `TreeView`, I got feedback from the a11y team that we actually need to move that click handler back up to the `<li>` for trees. It has to do with the `role="tree"` and child "role=treeitem"` markup and navigating the dom with a screen reader. If the child `<li>` is `treeitem` we want the screenreader to focus that item, so having a click handler elsewhere is confusing. In PVC that change requires another PCSS release and more JS refactor, so I have a [followup ticket](https://github.com/github/primer/issues/910) for that. 

So basically.. I think we should try and maintain a `<button>` for `NavList` and remove `<button>` from `TreeView`, meaning those two items have slightly different markup.

> restricted to just text

Both items can have all the same slots, but I've added two additional slots that `TreeView` utilizes: `leadingAction` and `trailingAction`. I created slots with CSS grid, so `ActionList-content` always has these slots available, but so far only `TreeView` and collapsible items have used the action slots. 

`grid-template-areas: 'leadingAction leadingVisual label trailingVisual trailingAction'`

Reply:
@siddharthkp and I tentatively decided to move forward with option 2:

```jsx
<NavList>
  <NavList.Item href="/branches">Branches</NavList.Item>
  <NavList.Item href="/tags">Tags</NavList.Item>
  <NavList.Item>
      Actions
      <NavList.LeadingVisual></ActionIcon/></NavList.LeadingVisual>
      <NavList.SubNav>
        <NavList.Item href="/actions">General</NavList.Item>
        <NavList.Item href="/actions/runners">Runners</NavList.Item>
      </NavList.SubNav>
  </NavList.Item/>
</NavList>
```

If anyone has strong opinions against this API, please let me know

Reply:
As @siddharthkp mentioned, the PVC `NavigationList` infers whether or not it should expand/collapse by determining if it has any sub-items, i.e. any item that has sub-items can be expanded/collapsed. That seemed like a reasonable assumption at the time. Will there ever be a need for a list of non-collapsible sub-items?

The API you've chosen makes sense, but I do wonder if simply using `<NavList.SubItem ...>` would be closer to the PVC API. For example:

```jsx
<NavList>
  <NavList.Item href="/branches">
    Branches
    <NavList.SubItem href="/actions">General</NavList.SubItem>
    <NavList.SubItem href="/actions/runners">Runners</NavList.SubItem>
  </NavList.Item/>
</NavList>
```

My understanding is navigation lists only ever have one nesting level.

Reply:
> My understanding is navigation lists only ever have one nesting level.

I think you're correct. We're working on a `TreeView` component to support more levels of nesting.

Reply:
> I do wonder if simply using <NavList.SubItem ...> would be closer to the PVC API

@siddharthkp and I are leaning towards the `<NavList.SubNav>` component because it will allow us to create a predictable and consistent API across `NavList`, `TreeView`, and `TreeGrid`. Would a similar API work for ViewComponents?

Reply:
@colebemis yeah that works for me 😄